{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e0a8466b",
   "metadata": {},
   "source": [
    "# 0. Introduction\n",
    "\n",
    "This tutorial will:\n",
    "\n",
    "- Use a small example dataset to show how to use wotplot to create and visualize binary dot plot matrices (**section 1**).\n",
    "\n",
    "- Demonstrate the use of wotplot with not-binary dot plot matrices (**section 2**).\n",
    "\n",
    "- Demonstrate a few techniques for creating fancy visualizations of dot plot matrices (**section 3**).\n",
    "\n",
    "- Use wotplot to create and visualize a large dot plot matrix comparing two _E. coli_ genomes (**section 4**).\n",
    "\n",
    "If you are unfamiliar with dot plot matrices, you may want to check out [the Wikipedia article on them](https://en.wikipedia.org/wiki/Dot_plot_(bioinformatics)). Chapter 6 of [_Bioinformatics Algorithms_](https://www.bioinformaticsalgorithms.org) also explains them well."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a629bc8",
   "metadata": {},
   "source": [
    "## 0.1. Import and set up a few things for later"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3803d7d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import wotplot\n",
    "from matplotlib import pyplot\n",
    "# the facecolor and transparent kargs force the saved figures to have a white background;\n",
    "# from https://stackoverflow.com/a/64585557\n",
    "savefig_kwargs = {\"bbox_inches\": \"tight\", \"facecolor\": \"white\", \"transparent\": False}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62009818",
   "metadata": {},
   "source": [
    "# 1. A small example"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6f43764",
   "metadata": {},
   "source": [
    "## 1.1. Define a small dataset\n",
    "\n",
    "Adapted from Figure 6.20 in Chapter 6 of _Bioinformatics Algorithms_ (Compeau & Pevzner), edition 2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7ae7a74c",
   "metadata": {},
   "outputs": [],
   "source": [
    "s1 = \"AGCAGGAGATAAACCTGT\"\n",
    "s2 = \"AGCAGGTTATCTACCTGT\"\n",
    "k = 3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f9834a8",
   "metadata": {},
   "source": [
    "## 1.2. Create a dot plot matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f5c74061",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "m = wotplot.DotPlotMatrix(s1, s2, k)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "995b9451",
   "metadata": {},
   "source": [
    "## 1.3. Inspect the dot plot matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1d2bc430",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "wotplot._matrix.DotPlotMatrix"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(m)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b0c8dbf3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DotPlotMatrix(mat=<16x16 sparse matrix of type '<class 'numpy.int64'>'\n",
       "\twith 18 stored elements in COOrdinate format>, k=3, yorder=\"BT\", binary=True)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a0d1750b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DotPlotMatrix(k = 3, binary, bottom â†’ top): 16 x 16\n"
     ]
    }
   ],
   "source": [
    "print(str(m))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29fdad46",
   "metadata": {},
   "source": [
    "The `DotPlotMatrix` object we just created, `m`, contains some extra information about the way it was created (e.g. the value of `k` we used). But the most interesting part of it is the actual matrix describing the dot plot! This matrix is stored in the `mat` attribute."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9379267f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<16x16 sparse matrix of type '<class 'numpy.int64'>'\n",
       "\twith 18 stored elements in COOrdinate format>"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "m.mat"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a91d9ecd",
   "metadata": {},
   "source": [
    "### 1.3.1. Why is this matrix 16x16? Shouldn't it be 18x18?\n",
    "\n",
    "You might have noticed that `s1` and `s2` are 18 nucleotides long. Why, then, does `m.mat` only have 16 rows and 16 columns?\n",
    "\n",
    "The reason for this is that the number of $k$-mers in an arbitrary string $s$ is slightly smaller than $|s|$: it's $|s| - k + 1$. For `s1` and `s2`, $|s| - k + 1 = 18 - 3 + 1 = 16$."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff4026da",
   "metadata": {},
   "source": [
    "### 1.3.2. What's the deal with sparse matrices?\n",
    "\n",
    "Most dot plot matrices are _sparse_: that is, most of their entries are zeroes. We exploit this by storing `m.mat` in a [sparse matrix format](https://en.wikipedia.org/wiki/Sparse_matrix), which allows us to only bother storing the non-zero entriesâ€”this drastically reduces the memory requirements when our input sequences are long (more than a few thousand nucleotides)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9af18e2b",
   "metadata": {},
   "source": [
    "#### Sidenote: what's the exact type of `m.mat`?\n",
    "Depending on the version of SciPy you have installed, `m.mat` will be of type [`scipy.sparse.coo_matrix`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.coo_matrix.html) (SciPy < 1.8) or type [`scipy.sparse.coo_array`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.coo_array.html) (SciPy â‰¥ 1.8). There shouldn't be much of a difference, at least for our use of these objects."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6c75fa5f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This notebook is using SciPy version 1.5.2.\n"
     ]
    }
   ],
   "source": [
    "import scipy\n",
    "print(f\"This notebook is using SciPy version {scipy.__version__}.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f29e862a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "scipy.sparse.coo.coo_matrix"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(m.mat)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6053417a",
   "metadata": {},
   "source": [
    "It's possible to convert `m.mat` from this sparse format to an equivalent \"dense\" format; this can make the matrix easier to work with, although it might require a large amount of memory if your input sequences were long. (For relatively small sequences like `s1` and `s2`, though, we should be fine.)\n",
    "\n",
    "Here's an example of creating a dense version of `m.mat`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b9e3031e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1],\n",
       "       [0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0],\n",
       "       [0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0],\n",
       "       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0],\n",
       "       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],\n",
       "       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],\n",
       "       [0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0],\n",
       "       [0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0],\n",
       "       [0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0],\n",
       "       [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0],\n",
       "       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0],\n",
       "       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0],\n",
       "       [0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0],\n",
       "       [0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0],\n",
       "       [0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],\n",
       "       [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]])"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "m.mat.toarray()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d808534",
   "metadata": {},
   "source": [
    "In the above matrix: `0` values represent cells where there are not any forward and/or reverse-complement $k$-mer matches, and `1` values represent cells where there is at least one such match.\n",
    "\n",
    "For more details about these values, you can run `help(wotplot.DotPlotMatrix)`."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3a8aa09",
   "metadata": {},
   "source": [
    "## 1.4. Visualize the dot plot matrix\n",
    "\n",
    "### 1.4.1. Available visualization functions\n",
    "Currently, we provide two functions for visualizing these matrices: `viz_imshow()` and `viz_spy()`. Both of these are essentially wrappers for matplotlib's [`imshow()`](https://matplotlib.org/stable/api/_as_gen/matplotlib.axes.Axes.imshow.html) and [`spy()`](https://matplotlib.org/stable/api/_as_gen/matplotlib.axes.Axes.spy.html) functions; you can even provide additional keyword arguments to `viz_imshow()` and `viz_spy()` which will be passed directly to `imshow()` / `spy()`. \n",
    "\n",
    "A brief summary of (in my opinion) the most important differences between these functions:\n",
    "\n",
    "- `imshow()`\n",
    "  - Draws zero and nonzero matrix cells as the same size, giving a \"perfect\" representation of the exact matrix.\n",
    "    - For small matrices (e.g. both sequences < 200 nt), this looks nice.\n",
    "    - For large matrices, the nonzero cells may be hard to see without enlarging the figure.\n",
    "  - Doesn't support sparse matrices.\n",
    "    - This means that `viz_imshow()` has to convert the sparse matrix to a dense format before calling `imshow()`. This will require a lot of memory if your matrix is large.\n",
    "\n",
    "\n",
    "- `spy()`\n",
    "  - Only draws nonzero matrix cells, meaning that the points representing each cell may cover other close-by cells in the matrix.\n",
    "    - You can increase / decrease nonzero cells' sizes as desired via the `markersize` parameter.\n",
    "    - For large matrices, this way of drawing things is actually nicer than the \"perfect\" representation offered by `imshow()` -- it makes nonzero cells much easier to see.\n",
    "  - Works with sparse matrices.\n",
    "    - This makes `viz_spy()` much more memory-efficient than `viz_imshow()`.\n",
    "\n",
    "In general, I recommend using `viz_imshow()` for small matrices (e.g. both sequences < 200 nt) and `viz_spy()` for large matrices.\n",
    "\n",
    "### 1.4.2. `viz_imshow()`\n",
    "\n",
    "First, let's use `viz_imshow()` -- our sequences here are very short, so the total cost of storing a dense 16 x 16 = 256-cell matrix in memory is small."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "75193267",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(<Figure size 432x288 with 1 Axes>,\n",
       " <AxesSubplot:xlabel='$s_1$ (18 nt) â†’', ylabel='$s_2$ (18 nt) â†’'>)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAQAAAAEACAYAAABccqhmAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAPKElEQVR4nO3dfbAddX3H8feHgIlAKQbCaG3LdQKOFWipKBAFW0WIddAEQWfsg0lpBR9SKdLC1DgD1I7o1HHQARXrIChTKKQZg+VBCSEVQkpCpJhMUBpCoogyCQF5yhCSfPvH7oXD9ebePXv2nHvO/X5eM2dOzu6e337v5O7n/vbht6uIwMxy2muiCzCzieMAMEvMAWCWmAPALDEHgFlie090AQcffHAMDQ1NdBlmA2HNmjV1v7o1ImaMnDjhATA0NMS999470WWYDQRJdb+6ebSJ3gUwS8wBYJaYA8AsMQeAWWIOALPEHABmiTkAzBJzAJgl5gAwS8wBYJaYA8AsMQeAWWITPhiol+oOpPB9E60b6vw+1v1d3NO63AMwS8wBYJaYA8AsMQeAWWIOALPEHABmiTkAzBJzAJgl5gAwS8wBYJaYA8AsMQeAWWKpBgM1PZCiG+uywTPIg8zcAzBLzAFglpgDwCwxB4BZYg4As8QcAGaJOQDMEnMAmCXmADBLzAFglpgDwCwxB4BZYg4As8RSjQasqx9GbVn3DfKovrrcAzBLzAFglpgDwCwxB4BZYg4As8QcAGaJOQDMEnMAmCXmADBLzAFglpgDwCwxB4BZYh4MZFYa5EE9dbkHYJaYA8AsMQeAWWIdBYCkqZJ+r6lizKy3ageApH2B24ANkk5rriQz65VaASBpf+BW4ARgH+A6h4DZ4Gk7AMqN/xbgWGAFEMCdOATMBk6dHsBJwJuBDwBLy2nvBe4Czm+oLjPrgbYvBIqIJZJmRsSjkt5UTtsu6VRgauMVmlnX1LoSMCIeHWXadmB7xxWZWc/4OgCzxBwAZok5AMwScwCYJeYAMEvMAWCWWKcBoPJlZgOoowCIiIsiwr0IswHljdcsMQeAWWIOALPEHABmiTkAzBJzAJgl5gAwS6zW/QAkvR44AjiE4pZgW4B1EfF/DdZmZl1WOQAk/QHwUeAM4NXDk8v3KJd5DLgeuCIiHmiwTjPrgnEDQNJM4AvAaRR3/LkTuAJ4CHicIgSmA4cBxwN/C/ydpMXABRGxsTulWyup/SuyMz4Lz16uSg9gPbAWmA8sjohnx1pY0n4UvYRzyu9O67BGM+uSKgHwgYi4sWqDZUBcDVwtaU7tysys68Y9C9DOxj/Kd5fU/a6ZdV+dB4NcKem4MeYfK+nKzsoys16ocx3AfGDmGPNfB8yrVY2Z9VQ3LgTaD3ihC+2aWcMqXQcg6feBoZZJb5D09lEWnQ58DNjQeWlm1m1VLwT6a+BCigt+AlhYvkYSsLtc3sz6XNUA+C6wiWIDvxL4BrByxDIBPAOsjoifN1SfmXVRpQCIiPuB+wEkHQr8Z0Ss62ZhZtZ9dZ4OfHE3CjGz3qs7GnAWsAA4HDiI37w1eETEWKcKzawPtB0Akj4MfIviVN+DwM+aLsraV2dgT50BRHXXZf2pTg9gIfBT4F0R8WjD9ZhZD9W5EOhQ4Gve+M0GX50AeASY2nQhZtZ7dQLg68BfSJrSdDFm1lt1jgGsAU4HVkm6HHgY2DVyoYj4YYe1mVmX1QmA21v+/U3K+wG2UDnNPQSzPlcnAHydv9kkUedKwKu7UYiZ9Z4fDGKWmAPALDEHgFliDgCzxBwAZonVGg5sk0PdUX0eRTh5dBwAKn4bhsq2NoT/l80GRuVdAEmXSNom6eeSziynnUTxkNANwE+ALZLO7k6pZta0qrcFnwdcQHHd/1bgCklbgWuBXwGXl23NBb4q6Vd+LJhZ/6u6C3AWcA9wYkTslHQJ8G3ggXLadgBJnwZ+BJwLOADM+lzVXYDXA9dGxM7y81XAAcDlwxs/QEQ8STFA6OjmSjSzbqkaAFOB51o+D/972yjLPg68spOizKw3qgbAJqD1icDD/37rKMu+DXisg5rMrEeqHgO4DrhY0q+BXwLnU9wReKakjwCLKMb/zwf+HPhO86WaWdOqBsCXgT8DPlV+fhL4G+BZ4G6K24RBcTOQbcA/N1eimXVL1UeDPVs+Dfg4ioN/q8oDfkg6juKo/+8A64FLI8LPCjAbAJWvBCyv8PufUaavBc5ssigz6w0PBjJLzAFglljjASDpLyUta7pd6x8RUeslqe2XdVc3egCHAn/ShXbNrGHeBTBLrOpowI1ttPnbNWsxsx6rehpwCHgCqPJE4H1rV2NmPVU1AB6muNvP7PEWlPQZ4OKOqjKznqh6DGAN8KaKy/qWYGYDomoA3AccJGmowrKbAT8Z2GwAVAqAiLgkIvaKiE0Vlr0mIt7RcWVm1nU+DWiWmAPALLFxA6C89Xctkt5V97tm1n1VegC3Slom6VRJU8ZbWNI+kk6T9N/AzZ2XaGbdUuU6gD8GvgTcSPHgj6XAKooHgmyjuAvQdOBw4HjgJOBA4Af47sDWopcPjaozkCjjQ63GDYCIWAecImkW8HFgDvAhfvN8v4CngMXA1yJidcO1mlnD2rkj0EpgZbkbcAzwRmAGRRBsAdYB90XE7m4UambNa/vhoBGxi2IXYFXz5ZhZL/k0oFliDgCzxBwAZok5AMwScwCYJeYAMEvMAWCWWCMBIOkMSZdKOlPS3iPm3dTEOsyseR0HgKQFwGUUNwP9R2CFpOkti5zY6TrMrDua6AEsAGZHxFnAUcD/AstaQsCPdzHrU00EwGsi4n6AiNgZEWcDy4A7JB2EbxJqE6BXjy4b9MeXNREAWyW9rnVCRHwKuKN8tT3ewMx6o4kAuB2YP3JiRPw9sByY1sA6zKwLmvjrvGBP7UTEJyV9sYF1mFkXtN0DkHSZpFuHP0fEDmCHpGMkvXrk8hHxsw5rNLMuqbML8A7gxY1axVGQ5RT3B3hE0iXNlGZm3VZnF+B3gbUtn98DvBW4BtgNnC9pRUT8VwP1mVkX1QmAAJ5r+TwH2ATMi4goz/+fAzgAzPpcnV2AzRT3Axzu/r8XWBIv3VL1+8Pzzay/1ekBXA+cJ+lm4DjgEF5+///dFLcJN7M+VycAvgJ8kOK+/6K49Hdpy/zDKO4SbGZ9rs5dgZ+W9BaKff8DgUUt3X/K6fc3U56ZdVOtC4HKc/83jJwu6WDgLl7eIzCzPtXodfoRsZVRLgsedH7MVA51/8/qDgjqh98R3xHILDEHgFliDgCzxBwAZok5AMwScwCYJeYAMEvMAWCWmAPALDEHgFliDgCzxBwAZok5AMwSS/XUnkEetWX9q5ejCJv+XXQPwCwxB4BZYg4As8QcAGaJOQDMEnMAmCXmADBLzAFglpgDwCwxB4BZYg4As8QcAGaJOQDMEhvY0YD9MJLKrBP98PvoHoBZYg4As8QcAGaJOQDMEnMAmCXmADBLzAFglpgDwCwxB4BZYg4As8QcAGaJOQDMEtNED0iQVKuAia7bbJBIWhMRbx453T0As8QcAGaJOQDMEnMAmCXmADBLzAFglpgDwCwxB4BZYg4As8QcAGaJOQDMEnMAmCXmADBLrB8eDbYV2Nzul+o8GswssUNHmzjhw4HNbOJ4F8AsMQeAWWIOALPEHABmiTkAzBJzAEwiko6UtFPSyRNdSxMkzZG0Q9LhE13LZOUAmFy+BKyIiNtaJ0r6J0k3SNooKSRtGqsRSftL+rSktZKelrRV0t2S5qvhCzAkHS3pIklDI+dFxBJgLfCFJtdpL3EATBKSZgEnU4TASJ8D3gk8BDwxTjt7AbcAnwVWA+cB/wJMAb4FfL65qgE4GrgQGNrD/C8Dp0k6ouH1Gg6AyeTjFFdV3jzKvJkRcVBEnAw8Ok47xwEnAF+JiDMj4hsRcSlwIvAwcHaDNVexGHgO+GiP15uCA6CHJB0gaaGkH0v6taSnJK2XdHmH7e4NzAWWRsQLI+dHxMY2mjugfH9ZUETEDoqAebZCPfPLXY13SvoHSQ9Jel7Sg5LmtSx3EUWvAuCO8jsh6aqW9T4D3Amc0cbPMF59cyX9VlPtDbJ+GAuQgqSpFL/Ih1L80q8H9gWOAjo9yHUMsD+wqsN2KNt4Eji/PFZwD0Wd88r1tPOX+HPAK4ErgOeBjwFXSdoQESso/rq/BjirXPaB8nsPjWhnJTBb0hsi4ic1fqYXSXotcB2wRtK7I+LpTtobdA6A3nkf8IfA7Ij4QcNtv7F8H7nhtC0inpD0PuCbwPUts54GTo+I77bR3FTgLWXvAUmLgI3AAoqDlT+WtJIiAG6LiOV7aGf45zoC6CgAIuIXkj4I3ADcUobAM520OcgcAL3zqvL9WElLI2J3g23PKN+3NdTeM8A64EbgbmA68Ang3yXNGXmWYQxfHd744cWN70Ha7/E8Xr4fMt6C5W5FFauBt1GEwCkRsb3NmiYFB0DvLKI4UPdZYIGk7wFLgJuHw6D8y/RJiiPjWyNiqGLbw0M6Oz5FJ+koio3+3Ij4esv0aylC4d8kzYyIXRWaG+3Yw+PsYWjqWGWV71WGrl7YZtvHUwRL20PSJwMfBOyRiNhGsQ/9buA/KE7ZfQ+4S9IrysWeAC4DFrbZ/JbyfXoDpZ4LTKPoIr8oIp4DbqLYeIcqtrWnkGg3qIZ/ri1jLgVEhMZ7AfsBy4CdwIciIuXGDw6AnoqIXRHx/Yg4B5gJXAPMAv6onH9bRFxH+3+N1pXvTVwx99ryfcoo8/Ye8d6EKn/VDyvf1425VAWS9qUI3rdTbPyLOm1zkDkAekDSjJFX0JVd6F0UG8AvOlzFfcBTFN3ZTq0v3+e3TpR0IDCHopeyoYH1DBs+ADdW7+V44LGI+GkD65tKccYk/cYPPgbQK18ETpC0hGLj2QuYDZwK/GtEjHdxzpgiYpekxcBcSVMj4vnW+ZL+ipf2u2cAr5D0mfLz5oj4TsvilwIfBj5fHg9YQbFxfoTilN0nKu7/V7Ua2A0slPQqiusMHo6Ie8ra96e4COnKJlZWnuWY1fBB2MEVEX51+UWxQd0EPEJxPvyXwO3AnD0sPxfY1OY6jqXoTZw+yrzl5bzRXstHWX4mcHVZ7wsUvYsfAu+vWMv8su0/3UMtm0ZMm0fR89hRfu+qEfMCOHKi/x8n48v3BOxDkuYCl0b1swDD37sV2C8iTuxGXRNB0o8oAuP9E13LZORjAH1E0hRJ04B9io+aVl5BWNV5wCxJp3Snwt4qg/BI4IIJLmXScg+gj0iaz0vXxg/b3G5PwKwqB4BZYt4FMEvMAWCWmAPALDEHgFliDgCzxBwAZok5AMwS+39aKaVVT++x4gAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "wotplot.viz_imshow(m)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5009e71",
   "metadata": {},
   "source": [
    "By default, the visualization will use black cells (â¬›) to represent matches (i.e. `1`s) and white cells (â¬œ) to represent no matches (i.e. `0`s)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1a4a53b",
   "metadata": {},
   "source": [
    "#### Sidenote: `Figure` and `Axes` objects\n",
    "\n",
    "What's the deal with the text above that says \"`(<Figure size 432x288 with 1 Axes>, <AxesSubplot:xlabel='$s_1$ (18 nt) â†’', ylabel='$s_2$ (18 nt) â†’'>)`\"?\n",
    "\n",
    "This text is showing up here because these two objects (a `matplotlib.figure.Figure` object and a `matplotlib.axes.Axes` object) are returned by `viz_imshow()` (and by `viz_spy()`).\n",
    "\n",
    "Internally, these functions create a `Figure` and an `Axes` object on which to draw the visualization (using [`pyplot.subplots()`](https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.subplots.html)), and return these objects for convenience's sake.\\* As we'll see later, it can be useful to have access to these objects -- for example, if we want to resize the plot, we can do this easily by calling the `set_size_inches()` method of the returned `Figure` object.\n",
    "\n",
    "\\* Note that this only applies if you leave the `ax` parameter of these functions set to the default of `None`. If you'd prefer, you can create these objects yourself and tell wotplot (using the `ax` parameter) to draw the dotplots on your specific `Axes` object -- this is useful when tiling multiple dot plots into a single figure, as is shown below in the section labelled \"Tiling multiple dot plots\"."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "191fbb06",
   "metadata": {},
   "source": [
    "### 1.4.3. For comparison's sake: `viz_spy()`\n",
    "\n",
    "The default `markersize` used by `viz_spy()` is `0.5`. I set this as the default because it's useful for plots of very long sequences; however, it's less useful for plots of tiny sequences like this one."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "70489125",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(<Figure size 432x288 with 1 Axes>,\n",
       " <AxesSubplot:xlabel='$s_1$ (18 nt) â†’', ylabel='$s_2$ (18 nt) â†’'>)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAQAAAAEECAYAAADH4+pwAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAP1UlEQVR4nO3dfbAddX3H8feHBBMJpRgIo7VtrhNwrEB7I4EQBduKEKcDJgg6Yx/MLVPBh1RKaWFqnAHGjujUcVIHVKyDUG2hkGYMDhAlBCqElDyYi8kEpSEPiiiTEJCnDJDk2z92Lzk53Ny7Z8+ex9/nNXPm5Ozu+e33TnY/Z59+u4oIzCxNh3W6ADPrHAeAWcIcAGYJcwCYJcwBYJYwB4BZwhwAZglzAJglzAFglrCJnS7g2GOPjYGBgU6XYdbX1q9fvysiptUP73gADAwMsG7duk6XYdbXJO0Ybbh3AcwS5gAwS5gDwCxhDgCzhDkAzPrE8PBww99xAJj1geHhYWbOnNlwCDgAzPrA4OAgGzZsYHBwsKHvOQDM+kSjKz84AMyS5gAwS5gDwCxhDgCzhDkAzBLmADBLmAPALGEOALOEOQDMEuYAMEuYAyBXpieVWSu0c1l0AFC+J5VZ1dq9LDoAKN+Tyqxq7V4WHQA5r/zWLdq5LDoAzBLmADBLmAPALGEOALOEOQDMEuYAMEuYA8AsYQ4As4Q5AMwS5gBokvsP2Fi6fflwADTBnYhsLL2wfDgAmuBORDaWXlg+HABN6ub/XOu8bl8+HABmCXMAmCXMAWCWMAeAWcIcAGYJcwCYJcwBYJYwB4BZwhwAZglzAJglzAFgVkA3d+hphgPAbBy90KuvLAeA2Th6oVdfWQ4AswL6ceUHB4BZ0hwAZglzAJglzAFgljAHgFnCHABmCXMAmCWsqQCQNEnS71VVjJm1V+kAkHQEcA+wRdL51ZVkZu1SKgAkHQksB84ADgdudQiY9Z6GAyBf+e8GTgNWAQE8gEPArOeU2QI4C5gFfBhYkQ87D3gQuKKiusysDSY2+oWIWCZpRkQ8Keld+bA9ks4FJlVeoZm1TMMBABART44ybA+wp+mKzKxtfB2AWcIcAGYJcwCYJcwBYJYwB4BZwhwAZglrNgCUv8ysBzUVABFxdUR4K8KsR3nlNUuYA8AsYQ4As4Q5AMwS5gAwS5gDwCxhDgCzhJW6H4CktwMnAseR3RJsJ7ApIv6vwtrMrMUKB4CkPwA+AVwIvHlkcP4e+TRPAbcBN0TEoxXWaWYtMG4ASJoBfAk4n+yOPw8ANwCPA0+ThcBU4HjgdOBvgL+VtBS4MiK2tqZ0M2tWkS2AzcBGYAhYGhEvjjWxpClkWwmX5t+d3GSNZtYiRQLgwxFxR9EG84C4GbhZ0rzSlZlZy417FqCRlX+U7y4r+10za70yDwa5UdLsMcafJunG5soys3Yocx3AEDBjjPFvAxaUqsbM2qoVFwJNAV5tQbtW0vDwcKdLsC5V6DoASb8PDNQMeoek944y6VTgk8CW5kuzKgwPDzNz5kw2bNjA4OBgp8uxLlP0QqC/Bq4iu+AngEX5q56A/fn01gUGBwe98tshFQ2A7wHbyVbwG4FvAqvrpgngBWBtRPyiovqsAl757VAKBUBEPAI8AiBpOvDfEbGplYWZWeuVeTrwNa0oxMzar2xvwDnAQuAE4Bhef2vwiIixThWaWRdoOAAkfQz4NtmpvseAn1ddlJm1R5ktgEXAz4D3R8STFddjZm1U5kKg6cDXvfKb9b4yAfAEMKnqQsys/coEwDeAv5A0oepizKy9yhwDWA9cAKyRdD2wDdhXP1FE/KjJ2sysxcoEwL01//4W+f0Aaygf5i2EHjY8POwrCBNQJgB8nX+fcweidJS5EvDmVhRi3cMdiNLhB4PYqLzyp8EBYJYwB4BZwhwAZglzAJglzAFglrBS9wOoJUlkNwydCGyJiPoLg8ysSxXeApB0raTdkn4h6aJ82FlkDwndAvwU2CnpktaUamZVK3pb8AXAlWTX/e8CbpC0C7gF+DVwfd7WfOBrkn7tx4KZdb+iuwAXAw8DZ0bEXknXAv8OPJoP2wMg6bPAj4HLAAeAWZcrugvwduCWiNibf74JOAq4fmTlB4iIZ8k6CA1WV6KZtUrRAJgEvFTzeeTfu0eZ9mngjc0UZb3LjyHrLUUDYDtQ+0TgkX+/e5Rp3wM81URN1qNGehE6BHpH0WMAtwLXSPoN8CvgCrI7As+Q9HFgCVn//yHgz4HvVF+qdTv3Iuw9KnLaXtIUYDnZrzvAs8B5wIvAQxy4R6DIdgtOjYhtRQqYNWtWrFu3rrGqzawhktZHxKz64UUfDfZi/jTg2WQH/9bkB/yQNJvsqP/vAJuBxRHhZwWY9YDCVwLmV/j97yjDNwIXVVmUmbWH+wKYJcwBYJawygNA0l9KWll1u2ZWvVZsAUwH/rgF7ZpZxbwLYJawor0BtzbQ5m+XrMXM2qzoacAB4BmgyBOBjyhdjZm1VdEA2EZ2t5+5400o6XPANU1VZUnxY8g6p+gxgPXAuwpO61uCWWHuQNRZRQNgA3CMpIEC0+4A/GRgK8QdiDqrUGegVnJnILPWO1RnIJ8GNEuYA8AsYeMGQH7r71Ikvb/sd82s9YpsASyXtFLSuZImjDexpMMlnS/pf4C7mi/RzFqlyHUAM4GvAHeQPfhjBbCG7IEgu8nuAjQVOAE4HTgLOBr4Ib47sFlXGzcAImITcI6kOcCngHnAR3n9+X4BzwFLga9HxNqKazWzijVyR6DVwOp8N+AU4J3ANLIg2AlsAjZExP5WFGpm1Wv44aARsY9sF2BN9eWYWTv5NKBZwhwAZglzAFhS3OnoYA4AS4Z7Hr6eA8CS4Z6Hr+cAsKR45T+YA8AsYZUEgKQLJS2WdJGkiXXj7qxiHmZWvaYDQNJC4Dqym4H+I7BK0tSaSc5sdh5m1hpVbAEsBOZGxMXAycAwsLImBFTBPMysBaoIgLdExCMAEbE3Ii4BVgL3SToG3yTUrGtVEQC7JL2tdkBE/D1wX/5quL+BmbVHFQFwLzBUPzAi/g64H5hcwTzMrAWq+HVeeKh2IuIzkr5cwTzMrAUa3gKQdJ2k5SOfI+IV4BVJp0h6c/30EfHzJms0sxYpswvwp8BrK7UkkW3qrwGekHRtNaWZWauVCYDfBTbWfP4z4N3AfwDfBa6QdG4FtZl1jX7tQFQmAAJ4qebzPGA7sCAihoA7gUubrsysS/RzL8IyAbCD7H6AI5v/5wHL4sAzxn4wMt6sH/RzL8IyZwFuAy6XdBcwGziOg+//v5/sNuFmfaMfV34oFwBfBT5Cdt9/kV36u6Jm/PFkdwk2sy5X5q7Az0s6lWzf/2hgSRz8iOF5wCPVlGdmrVTqQqD83P/t9cMlHQs8yMFbBGbWpSq9Tj8idjHKZcFm1p18RyCzhDkAzBLmADBLmAPALGEOALOEOQCa1I/Xh1t1un35cAA0oZ87iVjzemH5cAA0oZ87iVjzemH5cAA0qZv/c63zun35cACYJcwBYJYwB4BZwhwAZglzAJglzAFgljAHgFnCHABmCXMAmCXMAWCWMAdArps7bFha2rksOgDojV5bloZ2L4sOAHqj15alod3LogMg55XfukU7l0UHgFnCHABmCXMAmCXMAWCWMAeAWcIcAGYJcwCYJcwBYJYwB4BZwvoyAHxNv1kxfRcA7thjVlzfBYA79pgV13cBAO7YY1ZUXwaAmRXjADBLmAPALGEOALOEOQDMEuYAMEuYA8AsYQ4As4Q5AMwS5gAwS1hXB4A79Ji1VtcGgHv1mbVe1waAe/WZtV7XBgC4V59Zq3V1AJhZazkAzBLmADBLmAPALGGKiM4WIO0EdnS0CLP+Nz0iptUP7HgAmFnneBfALGEOALOEOQDMEuYAMEuYA6CPSDpJ0l5JZ3e6lipImifpFUkndLqWfuUA6C9fAVZFxD21AyX9k6TbJW2VFJK2j9WIpCMlfVbSRknPS9ol6SFJQ5JUZcGSBiVdLWmgflxELAM2Al+qcp52gAOgT0iaA5xNFgL1vgC8D3gceGacdg4D7gY+D6wFLgf+GZgAfBv4YnVVAzAIXAUMHGL8vwLnSzqx4vkaDoB+8ilgF3DXKONmRMQxEXE28OQ47cwGzgC+GhEXRcQ3I2IxcCawDbikwpqLWAq8BHyizfNNggOgjSQdJWmRpJ9I+o2k5yRtlnR9k+1OBOYDKyLi1frxEbG1geaOyt8PCoqIeIUsYF4sUM9QvqvxPkn/IOlxSS9LekzSgprpribbqgC4L/9OSLqpZr4vAA8AFzbwN4xX33xJv1VVe71sYqcLSIWkSWQL8nSyhX4zcARwMtDsQa5TgCOBNU22Q97Gs8AV+bGCh8nqXJDPp5Ff4i8AbwRuAF4GPgncJGlLRKwi+3V/C3BxPu2j+fcer2tnNTBX0jsi4qcl/qbXSHorcCuwXtIHIuL5ZtrrdQ6A9vkg8IfA3Ij4YcVtvzN/r19xGhYRz0j6IPAt4LaaUc8DF0TE9xpobhJwar71gKQlwFZgIdnByp9IWk0WAPdExP2HaGfk7zoRaCoAIuKXkj4C3A7cnYfAC8202cscAO3zpvz9NEkrImJ/hW2PdPLYXVF7LwCbgDuAh4CpwKeB/5Q0r/4swxi+NrLyw2sr32M0vsXzdP5+3HgT5rsVRawF3kMWAudExJ4Ga+oLDoD2WUJ2oO7zwEJJ3weWAXeNhEH+y/QZsiPjuyJioGDbIz26mj5FJ+lkspX+soj4Rs3wW8hC4d8kzYiIfQWaG+3Yw9Nku0ENlZW/F+m5dlWDbZ9OFixJ9kj1QcA2iYjdZPvQHwD+i+yU3feBByW9IZ/sGeA6YFGDze/M36dWUOplwGSyTeTXRMRLwJ1kK+9AwbYOFRKNBtXI37VzzKmAiNB4L2AKsBLYC3w0IpJc+cEB0FYRsS8ifhARlwIzgO8Cc4A/ysffExG30viv0ab8vYor5t6av08YZdzEuvcqFPlVPz5/3zTmVAVIOoIseN9LtvIvabbNXuYAaANJ0+qvoMs3ofeRrQC/bHIWG4DnyDZnm7U5fx+qHSjpaGAe2VbKlgrmM2LkANxYWy+nA09FxM8qmN8ksjMmya/84GMA7fJl4AxJy8hWnsOAucC5wL9ExHgX54wpIvZJWgrMlzQpIl6uHS/prziw3z0NeIOkz+Wfd0TEd2omXwx8DPhifjxgFdnK+XGyU3afLrj/X9RaYD+wSNKbyK4z2BYRD+e1H0l2EdKNVcwsP8sxp+KDsL0rIvxq8YtshboTeILsfPivgHuBeYeYfj6wvcF5nEa2NXHBKOPuz8eN9rp/lOlnADfn9b5KtnXxI+BDBWsZytv+k0PUsr1u2AKyLY9X8u/dVDcugJM6/f/Yjy/fEqwLSZoPLI7iZwFGvrccmBIRZ7airk6Q9GOywPhQp2vpRz4G0EUkTZA0GTg8+6jJ+RWERV0OzJF0TmsqbK88CE8CruxwKX3LWwBdRNIQB66NH7Gj0S0Bs6IcAGYJ8y6AWcIcAGYJcwCYJcwBYJYwB4BZwhwAZglzAJglzAFglrD/B+xa5aog3V7IAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "wotplot.viz_spy(m)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ea0237d",
   "metadata": {},
   "source": [
    "We can adjust the `markersize` when we call `viz_spy()`, which makes this visualization look essentially the same as what we'd get from `viz_imshow()`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "20c79d41",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(<Figure size 432x288 with 1 Axes>,\n",
       " <AxesSubplot:xlabel='$s_1$ (18 nt) â†’', ylabel='$s_2$ (18 nt) â†’'>)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAQAAAAEECAYAAADH4+pwAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAQPklEQVR4nO3de7BdZX3G8e9zAINAaQyXva1tOU7AsQItlQpEwVYRYjtogqAz9mJSWsFLKkUqTMUZoHZER8dBB6xYB6EyhULKGCwX5VohUAiRRjKgNEBQxJNJCMh1gOT8+sdam2wOJ+estfba1/f5zOzZOeu2f2ey13PedXnXq4jAzNI01u8CzKx/HABmCXMAmCXMAWCWMAeAWcIcAGYJcwCYJcwBYJYwB4BZwnbsdwF77rlnjI+P97sMs5G2evXqTRGx19TpfQ+A8fFx7r777n6XYTY0ms0mGzZsKLva3Okm+hDAbMhU2PlhO3/sHQBmCXMAmCXMAWCWMAeAWcIcAGYJcwCYJcwBYJYwB4BZwhwAZglzAJglzAFgljAHgNmQaTQaVVbbMt3E5AKg2WwiqdSr2Wz2u2wbQVW/ixMTE0REqRewZroakguAKj2pKva+MpvRIHwXkwsAM9vGAWCWMAeAWcIcAGYJcwCYJcwBYJYwB4BZwhwAZglzAJglzAFgljAHgFnCkguAKj2pGo2GOxHZjKp8P8bGyu9+FXsCblffxwbstYmJiUrrSSq9jjsRpaPK//Xk5GSrp17fJNcCMLNtHABmCXMAmCXMAWCWMAeAWcIcAGYJcwCYJcwBYJYwB4BZwhwAZglzAJglzAFQUNVORJaGYf1+JNcZqKqqnYhsuDSbzdIdexqNxtB+P9wCMGszCMN19ZIDwCxhDgCzhDkAzBLmADBLmAPALGEOALOEOQDMEuYAMEuYA8AsYQ4As4Q5AMwS5gAwazOsvfqqcm9AszbD2quvKrcAzBLWUQBImiPpd+oqxsx6q3IASNoFuB5YJ+nY+koys16pFACSdgOuAw4HdgIucwiYDZ/SAZDv/NcChwArgQBuxSFgNnSqtACOBP4I+CBwQz7tfcBtwGk11WVmPVD6MmBErJA0PyIek/TWfNrzko4B5tReoZl1TaX7ACLisWmmPQ8833FFZtYzvg/ALGEOALOEOQDMEuYAMEuYA8AsYQ4As4R1GgDKX2Y2hDoKgIg4KyLcijAbUt55zRLmADBLmAPALGEOALOEOQDMEuYAMEuYA8AsYZWeByDpTcD+wN5kjwTbCKyNiP+rsTYz67LCASDp94CPAccDzdbk/D3yZTYAlwMXRMT9NdZpZl0wawBImg98CTiW7Ik/twIXAA8Cj5OFwDxgX+Aw4G+Bv5N0JXB6RDzUndLNrFNFWgD3AfcCS4ErI+LZmRaWtCtZK+HkfN2dO6zRCmo2m2zYsKHUOo1GI7nhsGybIgHwwYi4qugG84C4GLhY0qLKlVlpZXf+quvY6Jj1KkCZnX+adVdUXdfMuq/KwCAXSjp0hvmHSLqws7LMrBeq3AewFJg/w/w3AksqVWNmPdWNG4F2BV7qwnbNrGaF7gOQ9LvAeNukN0t65zSLzgM+DqzrvDQz67aiNwL9NXAm2Q0/AZyRv6YSMJkvb2YDrmgAfA9YT7aDXwh8C7hjyjIBPAOsiohf1FSfmXVRoQCIiDXAGgBJ+wD/GRFru1mYmXVfldGBz+5GIWbWe1V7Ay4AlgH7AXvw6keDR0TMdKnQzAZA6QCQ9BHgO2SX+h4Afl53UWbWG1VaAGcAPwPeExGP1VyPdaDRaJS+t39sbAyp3Ngu7kA0OqoEwD7AZ7zzD54qO2XZnR/cgWiUVLkT8FFgTt2FmFnvVQmAbwJ/IWmHuosxs96qcgiwGjgOuEvS+cDDwNapC0XEjzqszcy6rEoA3Nj272+TPw+wjfJpbiGYDbgqAeD7/M1GRJU7AS/uRiFm1nseGMQsYQ4As4Q5AMwS5gAwS5gDwCxhDoDENRqNnqxjg6nS8wDaKetNMp5va11ETL0xyAZY1V59HoZsNBRuAUg6R9JmSb+QdEI+7UiyQULXAT8FNko6qTul2iDxMGSjoehjwZcAp5Pd978JuEDSJuBSYAI4P9/WYuAbkiY8LJjZ4Ct6CHAicCdwRERskXQO8G/A/fm05wEkfRb4MXAK4AAwG3BFDwHeBFwaEVvyny8CdgfOb+38ABHxJFkHoYPqK9HMuqVoAMwBnmv7ufXvzdMs+zjw2k6KMrPeKBoA64H2EYFb/377NMu+A/DZHrMhUPQcwGXA2ZJ+DfwKOI3sicDzJX0UWE7W/38p8OfAd+sv1czqVjQAvgb8KfDp/Ocngb8BngVuJ3tMGGQPA9kM/FN9JZpZtxQdGuzZfDTgQ8lO/t2Vn/BD0qFkZ/1/C7gPODciPFaA2RAofCdgfoff/0wz/V7ghDqLMrPecF8As4Q5AMwSVnsASPpLSTfVvV0bLFV6BLaGISvzajabXajeWjruDTiNfYA/7sJ2bYB4GLLR4EMAs4QV7Q34UIlt/mbFWsysx4oeAowDTwBFRgTepXI1ZtZTRQPgYbKn/SycbUFJnwPO7qgqM+uJoucAVgNvLbisHwlmNiSKBsA9wB6Sxgss+wjgkYHNhkChAIiIcyJiLCLWF1j2koh4V8eVmVnX+TKgWcIcAGYJmzUA8kd/VyLpPVXXNbPuK9ICuE7STZKOkbTDbAtL2knSsZL+G7im8xLNrFuK3Afwh8BXgavIBv64AbiLbECQzWRPAZoH7AccBhwJzAV+iJ8ObG0ajUal0YSse2YNgIhYCxwtaQHwCWAR8GFefb1fwFPAlcC/RMSqmmu1IdfLYcE8dFkxZZ4IdAdwR34YcDDwFmAvsiDYCKwF7omIyW4UalaGhy4rpnR34IjYSnYIcFf95ZhZL/kyoFnCHABmCXMAmCXMAWCWMAeAWcIcAGYJcwCYJayWAJB0vKRzJZ0gaccp866u4zPMrH4dB4CkZcB5ZA8D/QywUtK8tkWO6PQzzKw76mgBLAMWRsSJwIHA/wI3tYVA+dEgzKwn6giA10fEGoCI2BIRJwE3ATdL2gM/JNT6oEovwhR7HtYRAJskvbF9QkR8Grg5f3Vj+DGzGU1MTBARpV4TExM0m82kxi+sIwBuBJZOnRgRfw/cAuxcw2eY9URqvQjr+Ou8bHvbiYhPSfpKDZ9hZl1QugUg6TxJ17V+jogXgRclHSzpVW2hiPh5hzWaWZdUOQR4F/DyTq1szOdbyJ4P8Kikc+opzcy6rcohwG8D97b9/GfA24FLgEngNEkrI+K/aqjPzLqoSgAE8Fzbz4uA9cCSiIj8+v/JgAPAbMBVOQR4hOx5gK3m//uAFRHRut7/g9Z8MxtsVVoAlwOnSroGOBTYm1c+/3+S7DHhZjbgqgTA14EPkT33X2S3/t7QNn9fsqcEm9mAq/JU4KclvY3s2H8usLyt+U8+fU095ZlZN1W6ESi/9n/F1OmS9gRu45UtAjMbULXepx8Rm5jmtmAzG0x+IlBBqXUSSVXVXoTD+v1wT72CUuskkqqqYwNmV8TLGYTvh1sAZglzAJglzAFgljAHgFnCHABmCXMAmCXMAWCWMAeAWcIcAGYJcwCYJcwBYJYwB0BBHmrKZjKs34/kOgM1m83SnTAajUblTiKWhirfj1YPwjLq/i4m1wJwrz4bFIPwXUwuAMxsGweAWcIcAGYJcwCYJcwBYJYwB4BZwhwAZglzAJglzAFgljAHgFnCHABmCRvqAKgyHNPYWPlfeRB6bdnoGYQehEPdG7BKx4jJyUleOZq5WX8MQg/ToW4BmFlnHABmCXMAmCXMAWCWMAeAWcIcAGYJcwCYJcwBYJYwB4BZwhwAZglzAJglbKgDYBA6U5gNs4HoDOThusz6YyBaAIMwRJJZigYiAMysPxwAZglzAJglzAFgljAHgFnCHABmCXMAmCXMAWCWMAeAWcIcAGYJcwCYJUz9HiVH0kZgLuU7Jm0B1tRekNlo2ici9po6se8BYGb940MAs4Q5AMwS5gAwS5gDwCxhDoARIukASVskHdXvWuogaZGkFyXt1+9aRpUDYLR8FVgZEde3T5T0j5KukPSQpJC0fqaNSNpN0mcl3SvpaUmbJN0uaakk1VmwpIMknSVpfOq8iFgB3At8qc7PtG0cACNC0gLgKLIQmOoLwLuBB4EnZtnOGHAt8HlgFXAq8M/ADsB3gC/WVzUABwFnAuPbmf814FhJ+9f8uYYDYJR8AtgEXDPNvPkRsUdEHAU8Nst2DgUOB74eESdExLci4lzgCOBh4KQaay7iSuA54GM9/twkOAB6SNLuks6Q9BNJv5b0lKT7JJ3f4XZ3BBYDN0TES1PnR8RDJTa3e/7+iqCIiBfJAubZAvUszQ813i3pHyQ9KOkFSQ9IWtK23FlkrQqAm/N1QtJFbZ/7DHArcHyJ32G2+hZL+o26tjfMBmJcgBRImkP2Rd6H7Et/H7ALcCDQ6Umug4HdgLs63A75Np4ETsvPFdxJVueS/HPK/CX+AvBa4ALgBeDjwEWS1kXESrK/7q8HTsyXvT9f78Ep27kDWCjpzRHx0wq/08skvQG4DFgt6b0R8XQn2xt2DoDeeT/w+8DCiPhhzdt+S/4+dccpLSKekPR+4NvA5W2zngaOi4jvldjcHOBteesBScuBh4BlZCcrfyLpDrIAuD4ibtnOdlq/1/5ARwEQEb+U9CHgCuDaPASe6WSbw8wB0Duvy98PkXRDREzWuO1WJ4/NNW3vGWAtcBVwOzAP+CTw75IWTb3KMINvtHZ+eHnne4DyLZ7H8/e9Z1swP6woYhXwDrIQODoini9Z00hwAPTOcrITdZ8Hlkn6PrACuKYVBvlfpk+RnRnfFBHjBbfd6tHV8SU6SQeS7fSnRMQ326ZfShYK/yppfkRsLbC56c49PE52GFSqrPy9SM+1M0tu+zCyYHmk5HojwScBeyQiNpMdQ78X+A+yS3bfB26T9Jp8sSeA84AzSm5+Y/4+r4ZSTwF2JmsivywingOuJtt5xwtua3shUTaoWr/XxhmXAiJCs72AXYGbyLqUfzgiktz5wQHQUxGxNSJ+EBEnA/OBS4AFwB/k86+PiMso/9dobf5exx1zb8jfd5hm3o5T3utQ5K/6vvn72hmXKkDSLmTB+06ynX95p9scZg6AHpC019Q76PIm9FayHeCXHX7EPcBTZM3ZTt2Xvy9tnyhpLrCIrJWyrobPaWmdgJup9XIYsCEiflbD580hu2KS/M4PPgfQK18BDpe0gmznGQMWAscAX46I2W7OmVFEbJV0JbBY0pyIeKF9vqS/Yttx917AayR9Lv/5kYj4btvi5wIfAb6Ynw9YSbZzfpTskt0nCx7/F7UKmATOkPQ6svsMHo6IO/PadyO7CenCOj4sv8qxoOaTsMMrIvzq8otsh7oaeJTsevivgBuBRdtZfjGwvuRnHELWmjhumnm35POme90yzfLzgYvzel8ia138CPhAwVqW5tv+k+3Usn7KtCVkLY8X8/UumjIvgAP6/f84ii8/EmwASVoMnBvFrwK01rsO2DUijuhGXf0g6cdkgfGBftcyinwOYIBI2kHSzsBO2Y/aOb+DsKhTgQWSju5Ohb2VB+EBwOl9LmVkuQUwQCQtZdu98S2PlG0JmBXlADBLmA8BzBLmADBLmAPALGEOALOEOQDMEuYAMEuYA8AsYQ4As4T9P47bEtFJuycZAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "wotplot.viz_spy(m, markersize=13)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec251ecc",
   "metadata": {},
   "source": [
    "# 2. Distinguishing forward, reverse-complementary, and palindromic matches\n",
    "\n",
    "If we set `binary=False` when creating a `DotPlotMatrix` object, then wotplot will distinguish between forward, reverse-complementary, and palindromic matches. This can be useful for certain analyses; also, we can color these matches differently in the visualization!\n",
    "\n",
    "## 2.1. Using `binary=False` with the same dataset as above"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "dd5c9aa8",
   "metadata": {},
   "outputs": [],
   "source": [
    "n = wotplot.DotPlotMatrix(s1, s2, k, binary=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f3f163a",
   "metadata": {},
   "source": [
    "If we convert `n`'s matrix to a dense format, we see that now it doesn't just contain `0` and `1` values:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "b830209a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  1],\n",
       "       [ 0,  0, -1,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  1,  0],\n",
       "       [ 0,  0,  0, -1,  0,  0,  0,  0,  0,  0,  0,  0,  0,  1,  0,  0],\n",
       "       [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  1,  0,  0,  0],\n",
       "       [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0],\n",
       "       [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0],\n",
       "       [ 0,  0,  0,  0,  0,  0, -1,  0,  0,  0,  0,  0,  0,  0,  0,  0],\n",
       "       [ 0,  0,  0,  0,  0,  0,  0, -1,  0,  0,  0,  0,  0,  0,  0,  0],\n",
       "       [ 0,  0,  0,  0,  0,  0,  0,  0, -1,  0,  0,  0,  0,  0,  0,  0],\n",
       "       [ 0,  0,  0,  0,  0,  0,  0,  0,  0, -1,  0,  0,  0,  0,  0,  0],\n",
       "       [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0, -1,  0,  0,  0,  0],\n",
       "       [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0, -1,  0,  0,  0],\n",
       "       [ 0,  0,  0,  1,  0,  0,  0,  0,  0,  0,  0,  0,  0, -1,  0,  0],\n",
       "       [ 0,  0,  1,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0, -1,  0],\n",
       "       [ 0,  1,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0],\n",
       "       [ 1,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0]])"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n.mat.toarray()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aad2225c",
   "metadata": {},
   "source": [
    "In this not-binary matrix:\n",
    "\n",
    "- `1` values indicate forward matches,\n",
    "- `-1` values indicate reverse-complementary matches, and\n",
    "- `2` values indicate palindromic matches (there aren't any of those in this example).\n",
    "\n",
    "For reference, these values actually correspond to constants in wotplot's package (`wotplot.FWD`, `wotplot.REV`, and `wotplot.BOTH`, respectively; for binary matrices, `1` values correspond to `wotplot.MATCH`). If you're writing code that works with these matrices, I suggest using these constants instead of `1`, `-1`, etc.; this might improve readability a bit. (But it's not a huge deal.)\n",
    "\n",
    "Okay, now let's draw this matrix! `viz_imshow()` will automatically notice that this matrix was created using `binary=False`, and will draw it in color."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "def90ddc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAQAAAAEACAYAAABccqhmAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAPF0lEQVR4nO3decxldX3H8fcH0EGgVlmM1raMGTXWpbVaWVRsXbGNOuCW2EWmtIIL1aotpmKi1kY0GoNGVKxBqKZSpUS0IgoiVZEKjhaZ4FKWwQU1DKCyRQS+/eOcBx8fn5m5z73n3mf5vV/JzZ17zrm/830yz/k8v7P8zklVIalNuyx3AZKWjwEgNcwAkBpmAEgNMwCkhu223AXsu+++tX79+uUuQ1odNm8e72uwrar2Wzh92QNg/fr1fPWrX13uMqTVIRnva3D1YtPdBZAaZgBIDTMApIYZAFLDDACpYQaA1DADQGqYASA1zACQGmYASA0zAKSGGQBSw5Z9MNAsjTmOAm+bqKkY5xdy3F/G7azLHoDUMANAapgBIDXMAJAaZgBIDTMApIYZAFLDDACpYQaA1DADQGqYASA1zACQGtbUYKCBx1FMZV1ahVbxKDN7AFLDDACpYQaA1DADQGqYASA1zACQGmYASA0zAKSGGQBSwwwAqWEGgNQwA0BqmAEgNayp0YDjWgGDtjQLq3hU37jsAUgNMwCkhhkAUsMMAKlhBoDUMANAapgBIDXMAJAaZgBIDTMApIYZAFLDDACpYQ4Gkuas4kE947IHIDXMAJAaZgBIDZsoAJKsS/I7QxUjabbGDoAkewDnAJcnOXy4kiTNylgBkGQv4Gzg8cDdgNMMAWn1WXIA9Bv/p4EDgAuAAr6IISCtOuP0AJ4M/BHwPODcftozgS8Bxw5Ul6QZWPKFQFV1ZpINVXVNkkf1025N8gxg3eAVSpqasa4ErKprFpl2K3DrxBVJmhmvA5AaZgBIDTMApIYZAFLDDACpYQaA1LBJAyD9S9IqNFEAVNUbqspehLRKufFKDTMApIYZAFLDDACpYQaA1DADQGqYASA1bKz7ASR5MPAw4D50twS7FthSVf83YG2SpmzkAEjye8CLgecC952b3L9Xv8yPgY8CJ1XVNwesU9IU7DQAkmwA3gocTnfHny8CJwFXANfRhcDewAOBg4C/Bf4uyRnAa6rqyumUrvkyxgXZDT4KTwuM0gO4DLgU2AScUVU372jhJHvS9RJe0X939wlrlDQlowTA86rqE6M22AfEqcCpSTaOXZmkqdvpWYClbPyLfPfMcb8rafrGeTDIyUkO3MH8A5KcPFlZkmZhnOsANgEbdjD/AcARY1UjaaamcSHQnsAvptCupIGNdB1Akt8F1s+b9JAkT1hk0b2BlwCXT16apGkb9UKgvwZeT3fBTwHH9a+FAtzZLy9phRs1AD4ObKXbwE8G3g9cuGCZAm4CLq6q7w1Un6QpGikAquoS4BKAJPsD/1lVW6ZZmKTpG+fpwG+cRiGSZm/c0YAHA8cADwL24ddvDV5VtaNThZJWgCUHQJIXAh+kO9X3HeC7QxelpRtnYM84A4jGXZdWpnF6AMcB3waeUlXXDFyPpBka50Kg/YH3uvFLq984AfB9YN3QhUiavXEC4H3AXyTZdehiJM3WOMcANgPPAS5KciJwFXDHwoWq6gsT1iZpysYJgM/N+/cH6O8HOE/6afYQpBVunADwOn9pjRjnSsBTp1GIpNnzwSBSwwwAqWEGgNQwA0BqmAEgNWys4cBaG8Yd1ecowrVj4gBIErobhu4GXF7lf7O0Woy8C5Dk+CTXJ/lekiP7aU+me0jo5cC3gGuTHD2dUiUNbdTbgh8BvIbuuv9twElJtgEfAX4EnNi3dRjwniQ/8rFg0so36i7AUcBXgEOq6vYkxwP/Bnyzn3YrQJLXAl8DXgkYANIKN+ouwIOBj1TV7f3nU4B7AifObfwAVfUTugFCjxyuREnTMmoArANumfd57t/XL7LsdcA9JilK0myMGgBbgflPBJ7792MXWfZxwI8nqEnSjIx6DOA04I1Jfgr8EDiW7o7AG5K8CDidbvz/JuDPgQ8NX6qkoY0aAO8E/hR4Vf/5J8DfADcDX6a7TRh0NwO5Hvjn4UqUNC2jPhrs5v5pwAfSHfy7qD/gR5ID6Y76/xZwGXBCVfmsAGkVGPlKwP4Kv/9ZZPqlwJFDFiVpNhwMJDXMAJAaNngAJPnLJOcN3a5WjqrxXsnSX5quafQA9gf+eArtShqYuwBSw0YdDXjlEtr8zTFrkTRjo54GXA/cAIzyROA9xq5G0kyNGgBX0d3t59CdLZjkdcAbJ6pK0kyMegxgM/CoEZf1lmDSKjFqAHwd2CfJ+hGWvRrwycDSKjBSAFTV8VW1S1VtHWHZD1fVEyeuTNLUeRpQapgBIDVspwHQ3/p7LEmeMu53JU3fKD2As5Ocl+QZSXbd2cJJ7pbk8CT/DZw1eYmSpmWU6wD+EHgH8Am6B3+cC1xE90CQ6+nuArQ38CDgIODJwL2Az+LdgTXPLJ8ZNc5AohafabXTAKiqLcDTkhwMvBTYCLyAXz/fH+BnwBnAe6vq4oFrlTSwpdwR6ELgwn434NHAQ4H96ILgWmAL8PWqunMahUoa3pIfDlpVd9DtAlw0fDmSZsnTgFLDDACpYQaA1DADQGqYASA1zACQGmYASA0bJACSPDfJCUmOTLLbgnmfGmIdkoY3cQAkOQZ4N93NQP8RuCDJ3vMWOWTSdUiajiF6AMcAh1bVUcAjgP8FzpsXAj7fRVqhhgiA+1XVJQBVdXtVHQ2cB3w+yT54k1Atg1k9umy1P75siADYluQB8ydU1auAz/evJY83kDQbQwTA54BNCydW1d8D5wO7D7AOSVMwxF/nY7bXTlW9PMnbB1iHpClYcg8gybuTnD33uapuA25L8ugk9124fFV9d8IaJU3JOLsATwTu2qiThK6rfxHw/STHD1OapGkbZxfgt4FL533+M+CxwIeBO4Fjk1xQVf81QH2SpmicACjglnmfNwJbgSOqqvrz/68ADABphRtnF+BquvsBznX/nwmcWXXXPVU/Mzdf0so2Tg/go8Crk5wFHAjch1+9//+ddLcJl7TCjRMA7wKeT3ff/9Bd+nvuvPkPpLtLsKQVbpy7At+Y5DF0+/73Ak6f1/2nn37JMOVJmqaxLgTqz/1/bOH0JPsCX+JXewSSVqhBr9Ovqm0sclnwqudzppow7n/ZuAOCVsKviHcEkhpmAEgNMwCkhhkAUsMMAKlhBoDUMANAapgBIDXMAJAaZgBIDTMApIYZAFLDDACpYW09tWc1D9vSijXLUYRD/yraA5AaZgBIDTMApIYZAFLDDACpYQaA1DADQGqYASA1zACQGmYASA0zAKSGGQBSwwwAqWGrdzTgShhKJU1gJfw62gOQGmYASA0zAKSGGQBSwwwAqWEGgNQwA0BqmAEgNcwAkBpmAEgNMwCkhhkAUsOWfzDQ5s0O7JGWiT0AqWEGgNQwA0BqmAEgNcwAkBpmAEgNMwCkhhkAUsMMAKlhBoDUMANAapgBIDXMAJAallrmUXVJrgWuXtYipLVv/6rab+HEZQ8AScvHXQCpYQaA1DADQGqYASA1zACQGmYArCFJHp7k9iRPXe5ahpBkY5LbkjxouWtZqwyAteUdwAVVdc78iUn+KcnHklyZpJJs3VEjSfZK8toklya5Mcm2JF9OsikZ5xbOO1zXI5O8Icn6hfOq6kzgUuCtQ65Tv2QArBFJDgaeShcCC70ZeBJwBXDDTtrZBfg08CbgYuDVwL8AuwIfBN4yXNUAPBJ4PbB+O/PfCRye5GEDr1cYAGvJS4FtwFmLzNtQVftU1VOBa3bSzoHA44F3VdWRVfX+qjoBOAS4Cjh6wJpHcQZwC/DiGa+3CQbADCW5Z5LjknwjyU+T/CzJZUlOnLDd3YDDgHOr6hcL51fVlUto7p79+68ERVXdRhcwN49Qz6Z+V+NJSf4hyRVJfp7kO0mOmLfcG+h6FQCf779TSU6Zt96bgC8Cz13Cz7Cz+g5L8htDtbeaLf+TgRqRZB3dL/L+dL/0lwF7AI8AJj3I9WhgL+CiCduhb+MnwLH9sYKv0NV5RL+epfwlfjNwD+Ak4OfAS4BTklxeVRfQ/XW/H3BUv+w3++9dsaCdC4FDkzykqr41xs90lyT3B04DNid5elXdOEl7q50BMDvPAn4fOLSqPjtw2w/t3xduOEtWVTckeRbwAeCj82bdCDynqj6+hObWAY/pew8kOR24EjiG7mDlN5JcSBcA51TV+dtpZ+7nehgwUQBU1Q+SPB/4GPDpPgRumqTN1cwAmJ179+8HJDm3qu4csO25UV7XD9TeTcAW4BPAl4G9gZcB/55k48KzDDvwnrmNH+7a+L7D0ns81/Xv99nZgv1uxSguBh5HFwJPq6pbl1jTmmAAzM7pdAfq3gQck+STwJnAWXNh0P9lejndkfFtVbV+xLbnhnROfIouySPoNvpXVtX75k3/CF0o/GuSDVV1xwjNLXbs4Tq63aAlldW/jzJ09fVLbPsgumBpcki6BwFnpKqup9uHfjrwH3Sn7D4JfCnJ3fvFbgDeDRy3xOav7d/3HqDUVwK703WR71JVtwCfott414/Y1vZCYqlBNfdzXbvDpYCqys5ewJ7AecDtwAuqqsmNHwyAmaqqO6rqM1X1CmAD8GHgYOAP+vnnVNVpLP2v0Zb+fYgr5u7fv++6yLzdFrwPYZS/6g/s37fscKkRJNmDLnifQLfxnz5pm6uZATADSfZbeAVd34W+g24D+MGEq/g68DO67uykLuvfN82fmORewEa6XsrlA6xnztwBuB31Xg4CflxV3x5gfevozpg0v/GDxwBm5e3A45OcSbfx7AIcCjwDeFtV7ezinB2qqjuSnAEclmRdVf18/vwkf8Uv97v3A+6e5HX956ur6kPzFj8BeCHwlv54wAV0G+eL6E7ZvWzE/f9RXQzcCRyX5N501xlcVVVf6Wvfi+4ipJOHWFl/luPggQ/Crl5V5WvKL7oN6lPA9+nOh/8Q+BywcTvLHwZsXeI6DqDrTTxnkXnn9/MWe52/yPIbgFP7en9B17v4AvDsEWvZ1Lf9J9upZeuCaUfQ9Txu6793yoJ5BTx8uf8f1+LLewKuQEkOA06o0c8CzH3vbGDPqjpkGnUthyRfowuMZy93LWuRxwBWkCS7JtkduFv3Mbv3VxCO6tXAwUmeNp0KZ6sPwocDr1nmUtYsewArSJJN/PLa+DlXL7UnII3KAJAa5i6A1DADQGqYASA1zACQGmYASA0zAKSGGQBSw/4ftJxyCG9dF5MAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig, ax = wotplot.viz_imshow(n)\n",
    "# Since we include it in the README, we'll save this drawing to a file.\n",
    "# We can do this using the fig object returned by viz_imshow() (or by viz_spy()).\n",
    "fig.savefig(os.path.join(\"img\", \"small_example_dotplot.png\"), **savefig_kwargs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c6b9052",
   "metadata": {},
   "source": [
    "By default, the visualization will use red cells (ðŸŸ¥) to represent forward matches (i.e. `1`s), blue cells (ðŸŸ¦) to represent reverse-complementary matches (i.e. `-1`s), purple cells (ðŸŸª) to represent palindromic matches (i.e. `2`s), and white cells (â¬œ) to represent no matches (i.e. `0`s)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4bfc908",
   "metadata": {},
   "source": [
    "`viz_spy()` also works with `binary=False` matrices:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e1c70c9",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "wotplot.viz_spy(n, markersize=13)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3071ccf",
   "metadata": {},
   "source": [
    "## 2.2. Another dataset that includes palindromes\n",
    "\n",
    "Just for demonstration, let's see how palindromic $k$-mers look in these visualizations. Note that palindromic $k$-mers can only occur for even values of $k$ (see [this discussion](https://bioinformatics.stackexchange.com/q/156) for details).\n",
    "\n",
    "We'll create a \"self dot plot,\" in which both the horizontal and vertical axes of the dot plot matrix correspond to the same sequence (named `s3` below). \"Self dot plots\" like this one can be helpful for visualizing self-similarity; for another example of a self dot plot, see [this visualization](https://commons.wikimedia.org/wiki/File:Zinc-finger-dot-plot.png) from [this Wikipedia page](https://en.wikipedia.org/wiki/Dot_plot_(bioinformatics)).\n",
    "\n",
    "#### Sidenote: Aren't \"self dot plots\" like this symmetric?\n",
    "\n",
    "Yep! Exactly _how_ your self dot plot is symmetric will depend on the `yorder` of your `wotplot.DotPlotMatrix` (see below), but all self dot plots will be symmetric in some way. If you wanted, you could only visualize one \"triangle\" of a self dot plot matrix; wotplot's built-in visualization functions don't support this (yet), but see [ModDotPlot](https://github.com/marbl/ModDotPlot) for an example of doing this."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b419201",
   "metadata": {},
   "outputs": [],
   "source": [
    "s3 = \"AGCAGAAAGAGATAAACCTGT\"\n",
    "p = wotplot.DotPlotMatrix(s3, s3, 2, binary=False)\n",
    "fig, ax = wotplot.viz_imshow(p)\n",
    "# Adjust labels to make clear that both axes correspond to the same sequence\n",
    "s3_lbl = f\"$s_3$ ({len(s3)} nt) \\u2192\"\n",
    "ax.set_xlabel(s3_lbl, fontsize=18)\n",
    "ax.set_ylabel(s3_lbl, fontsize=18)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "347523cf",
   "metadata": {},
   "source": [
    "There are three palindromic $k = 2$-mers in `s3` (listed in the table below). These 2-mers are shown as purple cells in the visualization above.\n",
    "\n",
    "| 2-mer | Starting position in `s3` (0-indexed) |\n",
    "| --- | --- |\n",
    "| `GC` | 1 |\n",
    "| `AT` | 11 |\n",
    "| `TA` | 12 |"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb1d1a8d",
   "metadata": {},
   "source": [
    "## 2.3. Adjusting the color scheme\n",
    "\n",
    "If you'd prefer something other than the default red / blue / purple / white colors, you can adjust this with the `nbcmap` parameter of `viz_imshow()`. `nbcmap` should be a `dict` mapping the four possible cell values (`0`, `1`, `-1`, `2`) to decimal RGB triplet colors (see [Wikipedia](https://en.wikipedia.org/wiki/Web_colors#Extended_colors) for context).\n",
    "\n",
    "As an example, here we'll use `nbcmap` to create a \"dark mode\" version of the above dot plot:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6bd2bbe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# it's gamer time\n",
    "# no match = black; forward match = white; RC match = green; palindromic match = pink\n",
    "dark_mode_cmap = {0: [0, 0, 0], 1: [255, 255, 255], -1: [0, 255, 0], 2: [255, 0, 255]}\n",
    "fig, ax = wotplot.viz_imshow(p, nbcmap=dark_mode_cmap)\n",
    "ax.set_xlabel(s3_lbl, fontsize=18)\n",
    "ax.set_ylabel(s3_lbl, fontsize=18)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39af47f4",
   "metadata": {},
   "source": [
    "# 3. Other fancy visualization options"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b4138e7f",
   "metadata": {},
   "source": [
    "## 3.1. Tiling multiple dot plots; flipping the y-axis\n",
    "\n",
    "One of the main reasons I wrote this library was so that I could create figures containing grids of many dot plots using matplotlib. wotplot makes this process fairly painless!\n",
    "\n",
    "Here's an example of creating a single figure containing two side-by-side dot plots. For our second dot plot we'll reuse the original dataset from above but, this time, set `yorder=\"TB\"` -- this will have the effect of \"flipping\" the y-axis so that the second sequence begins at the top and ends at the bottom of the plot. This is analogous to how [Gepard](https://academic.oup.com/bioinformatics/article/23/8/1026/198110)'s dot plots look."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40a1ea34",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Second dot plot: same as the original, but different y-order\n",
    "m_flipped = wotplot.DotPlotMatrix(s1, s2, k, yorder=\"TB\")\n",
    "\n",
    "fig, (axLeft, axRight) = pyplot.subplots(1, 2)\n",
    "\n",
    "# Notice how we provide an argument to the ax parameters of these functions.\n",
    "# This way, we can create these Axes objects in advance (when we call pyplot.subplots()).\n",
    "wotplot.viz_imshow(m, title=\"Bottom-to-top\", ax=axLeft)\n",
    "wotplot.viz_imshow(m_flipped, title=\"Top-to-bottom\", ax=axRight)\n",
    "\n",
    "fig.set_size_inches(12, 5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "123d9ec4",
   "metadata": {},
   "source": [
    "## 3.2. Passing arbitrary keyword arguments to `imshow()` / `spy()`\n",
    "\n",
    "These functions have a lot of options available (see [the matplotlib docs for `imshow()`](https://matplotlib.org/stable/api/_as_gen/matplotlib.axes.Axes.imshow.html) and [the matplotlib docs for `spy()`](https://matplotlib.org/stable/api/_as_gen/matplotlib.axes.Axes.spy.html)), and we can make use of these options without too much effort:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a6c3f79",
   "metadata": {},
   "outputs": [],
   "source": [
    "# - Set cmap to \"viridis\" (by default, viz_imshow() sets cmap to \"gray_r\" for binary matrices, but we can\n",
    "#   override this here).\n",
    "# - Set aspect to 0.5 (stretches out the x-axis; this can be useful if you're creating a dot plot where\n",
    "#   the sequence used for the x-axis is smaller than the sequence used for the y-axis).\n",
    "wotplot.viz_imshow(m, cmap=\"viridis\", aspect=0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30be1fa7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# - Set marker to \"o\" (draw circles instead of squares for matching cells).\n",
    "# - Set alpha to 0.3 (add some transparency to the matching cells).\n",
    "wotplot.viz_spy(m, markersize=13, marker=\"o\", alpha=0.3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d426aa3",
   "metadata": {},
   "source": [
    "## 3.3. Adding ticks, tick labels, and a grid (a.k.a. activating Fancy Mode)\n",
    "We'll show this by reusing the above example.\n",
    "\n",
    "Adding ticks can be a little complicated, especially if we want the labels to be correct. So this code is a bit long, sorry -- I've tried to document it thoroughly to make it easier to read."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad05eb6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# While we're adding this other explanatory stuff, let's add a title for good measure.\n",
    "fig, ax = wotplot.viz_spy(m, markersize=13, marker=\"o\", alpha=0.3, title=f\"Dot plot ($k = {m.k}$)\")\n",
    "\n",
    "###### PART 1: TICKS ######\n",
    "# First, let's add ticks. We need to add ticks for ax.grid() to work.\n",
    "# We'll add both major and minor ticks -- the major ticks will be used to position the grid lines,\n",
    "# and the minor ticks will be used to position the tick labels. This allows us to center our labels\n",
    "# in between ticks, which is nice when visualizing a matrix. This approach is inspired by\n",
    "# https://matplotlib.org/stable/gallery/ticks/centered_ticklabels.html. (For context on major vs.\n",
    "# minor ticks, see https://matplotlib.org/stable/gallery/ticks/major_minor_demo.html.)\n",
    "\n",
    "# Our major ticks' positions will look like [-0.5, 0.5, 1.5, ..., 13.5, 14.5, 15.5]. We\n",
    "# position them like this because the cells of the matrix occur at integer coordinates.\n",
    "majticks = [t - 0.5 for t in range(0, 17)]\n",
    "\n",
    "# We'll add minor ticks in between each of the major ticks. Our minor ticks use a smaller\n",
    "# range than our major ticks because there are (n - 1) spaces between n consecutive points (e.g.\n",
    "# your hands probably have 5 fingers each, but 4 spaces-between-fingers).\n",
    "minticks = range(0, 16)\n",
    "\n",
    "# The x-axis is simple -- we'll just use these major and minor ticks.\n",
    "ax.set_xticks(majticks)\n",
    "ax.set_xticks(minticks, minor=True)\n",
    "\n",
    "# The y-axis is a bit more complicated. Keep in mind that, when creating the DotPlotMatrix \"m\"\n",
    "# above, we left the DotPlotMatrix \"yorder\" parameter at its default value of \"BT\" -- so the\n",
    "# y-axis goes from bottom to top.\n",
    "#\n",
    "# However, matplotlib doesn't know about this -- by default, it'll consider \"row 0\" to be the\n",
    "# topmost row in the matrix, then row 1 the second-from-the-top row, ...)\n",
    "# We can address this by adjusting the ticks on the y-axis (so that \"row 0\" is now the\n",
    "# bottommost row, etc.) -- we'll do this using the t2y() function defined below.\n",
    "def t2y(tl):\n",
    "    return [m.mat.shape[0] - t - 1 for t in tl]\n",
    "# Now, we can apply these corrected ticks to the y-axis.\n",
    "ax.set_yticks(t2y(majticks))\n",
    "ax.set_yticks(t2y(minticks), minor=True)\n",
    "\n",
    "###### PART 2: TICK LABELS ######\n",
    "# Next, let's add tick labels for the minor axis ticks. Adding one label for every minor tick\n",
    "# is a bit excessive, so let's just add labels for every other minor tick. (For dot plots of\n",
    "# long sequences, you'd probably want to add ticks and tick labels a lot farther apart -- e.g.\n",
    "# separated by 1 Mbp.)\n",
    "tick_labels = []\n",
    "# Even though Python uses 0-indexing, we'll use 1-indexing here to make this plot a bit easier\n",
    "# to interpret -- so we'll use range(1, 17) instead of range(0, 16).\n",
    "for i in range(1, 17):\n",
    "    if i % 2 == 1:\n",
    "        # Only add labels for odd-numbered positions.\n",
    "        tick_labels.append(f\"{i:,}\")\n",
    "    else:\n",
    "        # We still need to add something for the even-numbered positions (to match the\n",
    "        # number of ticks we added above), so we'll add an empty string for each of these ticks.\n",
    "        tick_labels.append(\"\")\n",
    "\n",
    "# Since we already reversed the ticks for the y-axis to account for how yorder=\"BT\", we can now\n",
    "# provide the tick labels for both the x- and y-axes in ascending order.\n",
    "ax.set_xticklabels(tick_labels, minor=True)\n",
    "ax.set_yticklabels(tick_labels, minor=True)\n",
    "\n",
    "# Show ticks and labels on the left and bottom axes only\n",
    "ax.tick_params(\n",
    "    labeltop=False, labelbottom=True, labelleft=True, labelright=False,\n",
    "    top=False, bottom=True, left=True, right=False, which=\"both\"\n",
    ")\n",
    "# Also, don't show minor ticks' lines -- we'll still show their labels,\n",
    "# but showing their tick lines (when we're already showing the major ticks' lines)\n",
    "# is unnecessary (in my opinion)\n",
    "ax.tick_params(bottom=False, left=False, which=\"minor\")\n",
    "\n",
    "###### PART 3: GRID LINES ######\n",
    "# This part is easy :)\n",
    "#\n",
    "# We'll only draw grid lines for the major ticks -- this creates the nice appearance of minor\n",
    "# ticks being lined up with their matrix rows / columns.\n",
    "#\n",
    "# Note that ax.grid() has a lot of other options you may want to play around with\n",
    "# (linestyle, color, ...), should you want to transcend Fancy Mode. (Editor's note:\n",
    "# Fancy Mode is not a real term and I made it up for this jupyter notebook.)\n",
    "ax.grid(which=\"major\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04bd6970",
   "metadata": {},
   "source": [
    "# 4. Creating dot plots of longer sequences\n",
    "\n",
    "As a final example, let's create a dot plot of two _E. coli_ strains' genomes. We'll use _E. coli_ K-12 [(from this assembly)](https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_000005845.2/) and _E. coli_ O157:H7 [(from this assembly)](https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_000008865.2/).\n",
    "\n",
    "I haven't included these FASTA files in the repository because they're both fairly large (both around ~5 MB), but -- for reference -- I downloaded them from the NCBI's website, moved them to a folder in this repository named `docs/data/`, and removed the two plasmid sequences from the O157:H7 assembly. (The K-12 assembly doesn't include any plasmid sequences; I guess this makes sense, given [the history of the K-12 strain](https://en.wikipedia.org/wiki/Escherichia_coli_in_molecular_biology#K-12)?)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b38c3ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Note that pyfastx (https://github.com/lmdu/pyfastx), the library I use here to\n",
    "# load these FASTA files' sequences into memory, isn't included as a dependency of\n",
    "# wotplot; you can load your sequences however you'd like.\n",
    "import pyfastx\n",
    "e1 = pyfastx.Fasta(os.path.join(\"data\", \"ecoli_k12.fna\"))\n",
    "e2 = pyfastx.Fasta(os.path.join(\"data\", \"ecoli_o157h7.fna\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c353c78",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract the sequences from these pyfastx.Fasta objects\n",
    "e1s = str(e1[0])\n",
    "e2s = str(e2[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f93d0f94",
   "metadata": {},
   "source": [
    "#### Sidenote: how do I select the $k$-mer size when dealing with big sequences?\n",
    "I don't have a single perfect answer. In general: smaller values of $k$ will show more details, but they'll increase the number of random matches between the two sequences, which will in turn increase the density of the dot plot matrix, which will in turn increase the memory footprint of the matrix. So you should consider the length of your sequences, their expected degree of similarity, what sort of \"signals\" you're looking for, and how much memory your system has when selecting the $k$-mer size.\n",
    "\n",
    "If you're really curious, here are some vague, informal suggestions that might be a good starting point. For tiny sequences (e.g. < 100 bp), any $k$ should be ok. For small-ish sequences (e.g. 100 bp â€“ 1 kbp), any $k > 2$ should be ok. For sequences in the range 1â€“100 kbp, I tend to use $k = 10$. For large sequences (e.g. > 100 kbp), I recommend using $k > 10$. (But none of this is written in stone, so please don't quote me on it...)\n",
    "\n",
    "#### Sidenote: I didn't read that sidenote and you're bad at writing documentation\n",
    "Wow, that's kind of harsh??? But don't worry, you didn't miss much.\n",
    "\n",
    "## 4.1. Creating a binary dot plot matrix for the two _E. coli_ strains\n",
    "\n",
    "We'll use a $k$-mer size of $k$ = 20. We'll also set `verbose=True` in order to get detailed logging output as the matrix is constructed -- this helps reassure us that wotplot isn't frozen."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c43f8b3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "em = wotplot.DotPlotMatrix(e1s, e2s, 20, verbose=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8902f8f0",
   "metadata": {},
   "source": [
    "Creating the matrix usually takes about 3 minutes (on my five-year old laptop with 8 GB of RAM). All things considered, not too shabby!\n",
    "\n",
    "Now let's visualize the matrix. We pretty much have to use `viz_spy()` here -- using `viz_imshow()` is not feasible, because creating a dense-format copy of this matrix would require us to have... wait, how many cells were in that matrix?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59645c8a",
   "metadata": {},
   "outputs": [],
   "source": [
    "em.mat.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "afd6818d",
   "metadata": {},
   "outputs": [],
   "source": [
    "f\"The matrix has {em.mat.shape[0] * em.mat.shape[1]:,} cells.\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d928ba5",
   "metadata": {},
   "source": [
    "Okay, so if we make the optimistic assumption that each cell in the matrix can be stored in a single bit (I don't think this is true even if the matrix is binary), then we'd still need ~3.19 terabytes (!!!) of memory to store the matrix in dense format. That's not happening on my laptop, so we'll have to use `viz_spy()`."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea043999",
   "metadata": {},
   "source": [
    "Again, we can use `verbose=True` for `viz_spy()` in order to get information about how long visualization is taking. (This should go by quickly, though -- visualization is fast, compared to the process of creating the matrix.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8bc2e3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = pyplot.subplots()\n",
    "wotplot.viz_spy(\n",
    "    em, markersize=0.01, title=f\"Comparison of two $E. coli$ genomes ($k$ = {em.k})\", ax=ax, verbose=True\n",
    ")\n",
    "ax.set_xlabel(f\"$E. coli$ K-12 substr. MG1655 ({len(e1s)/1e6:.2f} Mbp) \\u2192\")\n",
    "ax.set_ylabel(f\"$E. coli$ O157:H7 str. Sakai ({len(e2s)/1e6:.2f} Mbp) \\u2192\")\n",
    "fig.set_size_inches(8, 8)\n",
    "fig.savefig(os.path.join(\"img\", \"ecoli_example_dotplot.png\"), **savefig_kwargs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a66b1c8a",
   "metadata": {},
   "source": [
    "There we have it! You may want to play around with the `markersize` parameter a bit in order to see what looks best."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa442b30",
   "metadata": {},
   "source": [
    "## 4.2. Creating the two _E. coli_ strains' dot plot matrix with `binary=False`\n",
    "\n",
    "When we're working with long sequences like these, the process of dot plot matrix visualization is a bit faster if we only want a binary matrix.\n",
    "\n",
    "However, it is possible to create a not-binary matrix (distinguishing forward, reverse-complementary, and palindromic matches) from this massive dataset. Visualizing it just takes a few extra seconds."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7cf18a01",
   "metadata": {},
   "outputs": [],
   "source": [
    "em_nb = wotplot.DotPlotMatrix(e1s, e2s, 20, binary=False, verbose=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8633bd1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# em and em_nb should have the same dimensions!\n",
    "em.mat.shape == em_nb.mat.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7bf44264",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This code is the exact same as the above visualization of the binary E. coli matrix, \n",
    "fig, ax = pyplot.subplots()\n",
    "wotplot.viz_spy(\n",
    "    em_nb, markersize=0.01, title=f\"Comparison of two $E. coli$ genomes ($k$ = {em_nb.k})\", ax=ax, verbose=True\n",
    ")\n",
    "ax.set_xlabel(f\"$E. coli$ K-12 substr. MG1655 ({len(e1s)/1e6:.2f} Mbp) \\u2192\")\n",
    "ax.set_ylabel(f\"$E. coli$ O157:H7 str. Sakai ({len(e2s)/1e6:.2f} Mbp) \\u2192\")\n",
    "fig.set_size_inches(8, 8)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8446dd48",
   "metadata": {},
   "source": [
    "Visualizing this matrix took a little bit longer, but it looks pretty!\n",
    "\n",
    "#### Sidenote: generating both binary and not-binary visualizations of a matrix at once\n",
    "\n",
    "If you'd like, you can use the `force_binary` parameter of `viz_spy()` to draw the matrix as if it were binary. This way, you can generate both of the above versions of this dot plot without having to recompute the `DotPlotMatrix` object."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
